File: frontend/packages/utils/src/event-readiness.ts
/** Local event readiness — RF-03A (mirrors backend event_readiness). */

import type {
  EventContentClassification,
  EventReadinessCheck,
  EventReadinessFields,
  EventReadinessLevel,
  TerritoryEventHealthFields,
  TerritoryEventHealthLevel,
} from "@yunicity/types";

export const EVENT_READINESS_LABELS: Record<EventReadinessLevel, string> = {
  ready: "Prêt",
  partial: "Partiel",
  not_ready: "Non prêt",
};

export const EVENT_CONTENT_CLASSIFICATION_LABELS: Record<EventContentClassification, string> = {
  real: "Réel",
  partial: "Partiel",
  placeholder: "Placeholder",
};

export const TERRITORY_EVENT_HEALTH_LABELS: Record<TerritoryEventHealthLevel, string> = {
  healthy: "Agenda vivant",
  warning: "Agenda faible",
  critical: "Aucun événement à venir",
};

const PLACEHOLDER_DESCRIPTION_SNIPPET =
  "un moment pilote proposé dans le cadre du réseau partenaire yunicity";

const VAGUE_EVENT_TITLES = new Set([
  "afterwork découverte",
  "découverte culinaire",
  "atelier ressources locales",
  "conseils style & entretien",
  "événement local",
  "rencontre locale",
]);

export function eventReadinessLabel(level: EventReadinessLevel): string {
  return EVENT_READINESS_LABELS[level];
}

export function territoryEventHealthLabel(level: TerritoryEventHealthLevel): string {
  return TERRITORY_EVENT_HEALTH_LABELS[level];
}

export function isEventPlaceholder(input: {
  title: string;
  description?: string | null;
}): boolean {
  const description = (input.description ?? "").trim().toLowerCase();
  const title = (input.title ?? "").trim().toLowerCase();
  if (description.includes(PLACEHOLDER_DESCRIPTION_SNIPPET)) return true;
  if (VAGUE_EVENT_TITLES.has(title) && description.length < 40) return true;
  return false;
}

export interface EventReadinessInput {
  title: string;
  description?: string | null;
  starts_at: string;
  ends_at?: string | null;
  location_name: string;
  address?: string | null;
  visibility: string;
  moderation_status: string;
  is_cancelled: boolean;
}

function classifyEventContent(input: {
  title: string;
  description?: string | null;
  location_name: string;
  placeholder: boolean;
}): EventContentClassification {
  if (input.placeholder) return "placeholder";
  const titleOk = input.title.trim().length >= 3;
  const descriptionOk = (input.description ?? "").trim().length >= 25;
  const locationOk = input.location_name.trim().length >= 2;
  if (titleOk && descriptionOk && locationOk) return "real";
  if (titleOk || descriptionOk) return "partial";
  return "placeholder";
}

export function eventReadiness(
  input: EventReadinessInput,
  now = new Date(),
): EventReadinessFields {
  const placeholder = isEventPlaceholder(input);
  const classification = classifyEventContent({
    title: input.title,
    description: input.description,
    location_name: input.location_name,
    placeholder,
  });

  const titleOk = input.title.trim().length >= 3;
  const descriptionText = (input.description ?? "").trim();
  const descriptionOk =
    !placeholder &&
    descriptionText.length >= 25 &&
    !descriptionText.toLowerCase().includes(PLACEHOLDER_DESCRIPTION_SNIPPET);
  const startsAt = new Date(input.starts_at);
  const dateOk = !Number.isNaN(startsAt.getTime());
  const dateFuture = dateOk && startsAt.getTime() >= now.getTime();
  const locationOk = input.location_name.trim().length >= 2;
  const visibilityOk = input.visibility === "public";
  const publicationOk = input.moderation_status === "approved" && !input.is_cancelled;

  const checks: EventReadinessCheck[] = [
    {
      key: "title_defined",
      label: "Titre défini",
      passed: titleOk,
      severity: titleOk ? "ok" : "error",
    },
    {
      key: "description_defined",
      label: "Description définie",
      passed: descriptionOk,
      severity: descriptionOk ? "ok" : "error",
    },
    {
      key: "date_defined",
      label: "Date définie",
      passed: dateOk,
      severity: dateOk ? "ok" : "error",
    },
    {
      key: "location_defined",
      label: "Lieu défini",
      passed: locationOk,
      severity: locationOk ? "ok" : "error",
    },
    {
      key: "visibility_enabled",
      label: "Visibilité publique",
      passed: visibilityOk,
      severity: visibilityOk ? "ok" : "warning",
    },
    {
      key: "publication_enabled",
      label: "Publication activée",
      passed: publicationOk,
      severity: publicationOk ? "ok" : "warning",
    },
    {
      key: "date_upcoming",
      label: "Événement à venir",
      passed: dateFuture,
      severity: dateFuture ? "ok" : "warning",
    },
    {
      key: "not_placeholder",
      label: "Contenu réel (non placeholder)",
      passed: !placeholder,
      severity: placeholder ? "error" : "ok",
    },
  ];

  const coreReady = titleOk && descriptionOk && locationOk && !placeholder && dateOk;
  let readiness: EventReadinessLevel;
  if (coreReady && publicationOk && visibilityOk && dateFuture) {
    readiness = "ready";
  } else if (placeholder || !titleOk || input.is_cancelled) {
    readiness = "not_ready";
  } else if (coreReady || (titleOk && locationOk)) {
    readiness = "partial";
  } else {
    readiness = "not_ready";
  }

  const contributes =
    readiness === "ready" && publicationOk && visibilityOk && dateFuture && !placeholder;

  return {
    readiness,
    classification,
    contributes_to_territory: contributes,
    territory_contribution_label: contributes
      ? "Cet événement contribue à maintenir l'agenda actif."
      : "Cet événement est incomplet et n'améliore pas la vitalité du territoire.",
    checks,
  };
}

export function territoryEventHealth(upcomingPublishedCount: number): TerritoryEventHealthFields {
  const count = Math.max(upcomingPublishedCount, 0);
  let status: TerritoryEventHealthLevel;
  if (count >= 5) status = "healthy";
  else if (count >= 1) status = "warning";
  else status = "critical";

  const label = territoryEventHealthLabel(status);
  const signal_emoji = status === "healthy" ? "🟢" : status === "warning" ? "🟡" : "🔴";

  return {
    status,
    upcoming_published_count: count,
    label,
    signal_emoji,
  };
}

File: backend/app/core/event_readiness.py
"""Local event territorial readiness — RF-03A (no ticketing / YM spend)."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime
from enum import StrEnum
from typing import Literal

from app.core.local_event_constants import (
    LocalEventModerationStatus,
    LocalEventVisibility,
)

PLACEHOLDER_DESCRIPTION_SNIPPET = (
    "un moment pilote proposé dans le cadre du réseau partenaire yunicity"
)
VAGUE_EVENT_TITLES: frozenset[str] = frozenset(
    {
        "afterwork découverte",
        "découverte culinaire",
        "atelier ressources locales",
        "conseils style & entretien",
        "événement local",
        "rencontre locale",
    }
)

CheckSeverity = Literal["ok", "warning", "error"]


class EventReadinessStatus(StrEnum):
    READY = "ready"
    PARTIAL = "partial"
    NOT_READY = "not_ready"


class EventContentClassification(StrEnum):
    """Audit taxonomy — REAL / PARTIAL / PLACEHOLDER."""

    REAL = "real"
    PARTIAL = "partial"
    PLACEHOLDER = "placeholder"


@dataclass(frozen=True)
class EventReadinessCheck:
    key: str
    label: str
    passed: bool
    severity: CheckSeverity


@dataclass(frozen=True)
class EventReadinessResult:
    status: EventReadinessStatus
    classification: EventContentClassification
    contributes_to_territory: bool
    territory_contribution_label: str
    checks: tuple[EventReadinessCheck, ...]


@dataclass(frozen=True)
class EventReadinessInput:
    title: str
    description: str | None
    starts_at: datetime
    ends_at: datetime | None
    location_name: str
    address: str | None
    visibility: str
    moderation_status: str
    is_cancelled: bool


def _text(value: str | None) -> str:
    return (value or "").strip()


def is_event_placeholder(
    *,
    title: str,
    description: str | None,
) -> bool:
    description_lower = _text(description).lower()
    title_lower = _text(title).lower()

    if PLACEHOLDER_DESCRIPTION_SNIPPET in description_lower:
        return True
    if title_lower in VAGUE_EVENT_TITLES and len(_text(description)) < 40:
        return True
    return False


def classify_event_content(
    *,
    title: str,
    description: str | None,
    location_name: str,
    is_placeholder: bool,
) -> EventContentClassification:
    if is_placeholder:
        return EventContentClassification.PLACEHOLDER
    title_ok = len(_text(title)) >= 3
    description_ok = len(_text(description)) >= 25
    location_ok = len(_text(location_name)) >= 2
    if title_ok and description_ok and location_ok:
        return EventContentClassification.REAL
    if title_ok or description_ok:
        return EventContentClassification.PARTIAL
    return EventContentClassification.PLACEHOLDER


def _description_defined(*, description: str | None, is_placeholder: bool) -> bool:
    if is_placeholder:
        return False
    text = _text(description)
    return len(text) >= 25 and PLACEHOLDER_DESCRIPTION_SNIPPET not in text.lower()


def _date_future(*, starts_at: datetime, now: datetime) -> bool:
    return starts_at >= now


def event_readiness(
    data: EventReadinessInput,
    *,
    now: datetime | None = None,
) -> EventReadinessResult:
    moment = now or datetime.now(UTC)
    is_placeholder = is_event_placeholder(title=data.title, description=data.description)
    classification = classify_event_content(
        title=data.title,
        description=data.description,
        location_name=data.location_name,
        is_placeholder=is_placeholder,
    )

    title_ok = len(_text(data.title)) >= 3
    description_ok = _description_defined(
        description=data.description,
        is_placeholder=is_placeholder,
    )
    date_ok = data.starts_at is not None
    date_future = _date_future(starts_at=data.starts_at, now=moment)
    location_ok = len(_text(data.location_name)) >= 2
    visibility_ok = data.visibility == LocalEventVisibility.PUBLIC.value
    publication_ok = (
        data.moderation_status == LocalEventModerationStatus.APPROVED.value
        and not data.is_cancelled
    )

    checks = (
        EventReadinessCheck(
            key="title_defined",
            label="Titre défini",
            passed=title_ok,
            severity="ok" if title_ok else "error",
        ),
        EventReadinessCheck(
            key="description_defined",
            label="Description définie",
            passed=description_ok,
            severity="ok" if description_ok else "error",
        ),
        EventReadinessCheck(
            key="date_defined",
            label="Date définie",
            passed=date_ok,
            severity="ok" if date_ok else "error",
        ),
        EventReadinessCheck(
            key="location_defined",
            label="Lieu défini",
            passed=location_ok,
            severity="ok" if location_ok else "error",
        ),
        EventReadinessCheck(
            key="visibility_enabled",
            label="Visibilité publique",
            passed=visibility_ok,
            severity="ok" if visibility_ok else "warning",
        ),
        EventReadinessCheck(
            key="publication_enabled",
            label="Publication activée",
            passed=publication_ok,
            severity="ok" if publication_ok else "warning",
        ),
        EventReadinessCheck(
            key="date_upcoming",
            label="Événement à venir",
            passed=date_future,
            severity="ok" if date_future else "warning",
        ),
        EventReadinessCheck(
            key="not_placeholder",
            label="Contenu réel (non placeholder)",
            passed=not is_placeholder,
            severity="ok" if not is_placeholder else "error",
        ),
    )

    core_ready = (
        title_ok
        and description_ok
        and location_ok
        and not is_placeholder
        and date_ok
    )

    if core_ready and publication_ok and visibility_ok and date_future:
        readiness_status = EventReadinessStatus.READY
    elif is_placeholder or not title_ok or data.is_cancelled:
        readiness_status = EventReadinessStatus.NOT_READY
    elif core_ready or (title_ok and location_ok):
        readiness_status = EventReadinessStatus.PARTIAL
    else:
        readiness_status = EventReadinessStatus.NOT_READY

    contributes = (
        readiness_status == EventReadinessStatus.READY
        and publication_ok
        and visibility_ok
        and date_future
        and not is_placeholder
    )
    if contributes:
        contribution_label = (
            "Cet événement contribue à maintenir l'agenda actif."
        )
    else:
        contribution_label = (
            "Cet événement est incomplet et n'améliore pas la vitalité du territoire."
        )

    return EventReadinessResult(
        status=readiness_status,
        classification=classification,
        contributes_to_territory=contributes,
        territory_contribution_label=contribution_label,
        checks=checks,
    )
